Fix liquibase checksum error.

SidMorad opened the pull request:
This option is recommended for when your table has production data (see
https://stackoverflow.com/a/63007543/2179502 for details).

Tonterias commented:
Let me try it and i will let you know if I could fix it. Thanks a lot!!!!!!!!!

Tonterias commented:
Please SidMorad, let me know if I'm wrong. 

The changes you added in the PR were not done by Jhipster automatically. I will have to modify the changes that Jhipster does automatically and include yours. Am I right?

Thanks again!

Tonterias commented:
Hi again,


I tried on a new project https://github.com/Tonterias/liquimvntwoentitiesbis because I was having problems with the changelog table at Postgres, but when I did your changes it gave me this error:

    Table one field two error in Postgres 2020-09-19 12:24:44.462 ERROR 19400 --- [itiesbis-task-1] i.g.j.c.liquibase.AsyncSpringLiquibase   : Liquibase could not start correctly, your database is NOT ready: Validation Failed:

Then I updated the md5sum field to null and it gave this one:

2020-09-19 12:34:26.722 ERROR 10116 --- [itiesbis-task-1] liquibase.changelog.ChangeSet            : Change Set config/liquibase/changelog/20200919000000_alter_entity_One.xml::20200919000000::manual failed.  Error: ERROR: no existe la relaci¾n ½public.one╗ [Failed SQL: (0) ALTER TABLE public.one ADD fieldtwo VARCHAR(255)]
2020-09-19 12:34:26.727 ERROR 10116 --- [itiesbis-task-1] i.g.j.c.liquibase.AsyncSpringLiquibase   : Liquibase could not start correctly, your database is NOT ready: Migration failed for change set config/liquibase/changelog/20200919000000_alter_entity_One.xml::20200919000000::manual:
     Reason: liquibase.exception.DatabaseException: ERROR: no existe la relaci¾n ½public.one╗ [Failed SQL: (0) ALTER TABLE public.one ADD fieldtwo VARCHAR(255)]

So I'm still stuck on this. When I find the solution I'm going to make a video on YouTube. I can't beleive I'm the only one having problems with this.

Thanks again

SidMorad commented:
> The changes you added in the PR were not done by Jhipster automatically. I will have to modify the changes that Jhipster does automatically and include yours. Am I right?

Yes, JHipster doesn't support incremental changes AFAIK. Therefore you can still use `jhipster entity One` command and add fields to your entity, but as my commit shows you have to rollback liquibase changes and add new changeset(s) to achieve your goal.

> Table one field two error in Postgres 2020-09-19 12:24:44.462 ERROR 19400 --- [itiesbis-task-1] i.g.j.c.liquibase.AsyncSpringLiquibase   : Liquibase could not start correctly, your database is NOT ready: Validation Failed:

Here was my testing scenario:

1. I did checkout your first commit
2. I did used postgresql.yml to create development database(`docker-compose -f src/main/docker/postgresql.yml up -d`)
3. I did run `./mvnw`
4. I did check if database created(`docker exec -it container-id-xyz psql -U liquimvntwoentities`)
5. I kind of ignored your second commit ;-)
6. Then I moved forward to your last commit and I was executing `./mvnw` after each `git checkout`.
7. I did face your error.
8. I fixed your error (this pull-request)
9. `./mvnw` shows no error and `\d one` command shows second column added to the table one.

I think you will have same result as me if you follow above steps.
One thing that I see is different in our setup is you are using native postgresql with different username/password, and I did use postgresql as docker container with default configs. So that may be root of these new errors.

Tonterias commented:
Hi SidMorad,

First, thanks a lot and I have to add that I think you would do a good job as a Jhipster team member. They would welcome you!

Sorry for the delay, I could not work with it these days. And now that I can, I'm still getting stuck :-(

5.- I kind of ignored your second commit ;-)  That made me sad ;-) I felt so proud of that one!
8.- I fixed your error (this pull-request). Where is it? I can`t find it! 8-0

I'm not a programmer, but I think I'm not alone in having a hard time with this liquibase updates and I wish there would be a video on YouTube on how to do it. I will do it if i learn how to fix it!

Thanks again.

SidMorad commented:
Hi Mike,

> First, thanks a lot and I have to add that I think you would do a good job as a Jhipster team member. They would welcome you!

Thanks, I accept this as compliment ;-)

> 8.- I fixed your error (this pull-request). Do you mean this one SidMorad:fix-liquibase-checksum? SidMorad@14361d4

Yes.

I suggest you to retry all steps in a new database(preferably new docker container) and when you reach the-checksum-issue then you apply [my changes](https://github.com/Tonterias/liquimvntwoentities/pull/1/files).

I am curious to know what cause different results? so please update me with your progress.

Thanks.

Tonterias commented:
Thanks SidMorad

I see what you have done: 

1) you undo the modifications that were actually done in (src/main/resources/config/liquibase/changelog/20200917163902_added_entity_One.xml) 

2) and you create a new file by hand to include those modifications (src/main/resources/config/liquibase/changelog/20200919000000_alter_entity_One.xml) and 

3) finally, you add the modifications to the master file so they get executed (src/main/resources/config/liquibase/master.xml). 

4) As a side order you delete the new data. because the data at production should be added and not created from the file, right? (src/main/resources/config/liquibase/fake-data/one.csv)

Now, my question is: it looks like the automatic process only works for development as the data is deleted and then created, right? But it would make sense to have your changes (made as in your process) automatically done both in dev and in production (just the same in both environments and just leaving aside the CSV file).

Let me say it again. If I do it your way I can use it both in Prod and Dev (even keeping or discarding the data), but when I use the automatic process it only works on Dev (for sure it will never working on Prod when I keep my data).

Don't you think we should say that Jhipster team should modify the actual process and do it your way? I would definetely say it if you help me on justifing it.

Have a nice day!

SidMorad commented:
>4.  As a side order you delete the new data. because the data at production should be added and not created from the file, right? (src/main/resources/config/liquibase/fake-data/one.csv)

I did rollback `one.csv` changes because otherwise we will get Liquibase error. In another words `*.csv` files are similar to change-sets and Liquibase keep their hashes after they did apply to a database. (Please correct me if I am wrong)

> Now, my question is: it looks like the automatic process only works for development as the data is deleted and then created, right?

Yes. e.g. In-memory database.

> But it would make sense to have your changes (made as in your process) automatically done both in dev and in production (just the same in both environments and just leaving aside the CSV file).

Yes, I agree. In real world scenarios, I think domain models(entities and value object) are not perfect at start and they go through lots of modifications. for example in my current project which we did start with JHipster 5.1.0, I just count our `config/liquibase/changelog/*added_field*.xml` files and the result is `107`, which means I did what I did in this pull request at least 107 times!

> Don't you think we should say that Jhipster team should modify the actual process and do it your way? I would definetely say it if you help me on justifing it.

It seems you want to make my life easier! ;-) I definitely support you to push this forward. I think many projects will not have perfect models at start and I would say maybe less than 10% of projects could shape perfect models at start then use JHipster magic to finish it! and they must delay going to production stage, which is not common these days. Iterative development and evolutionary designs are more common, and people expect to see results as your team progress forward.
So our models are not perfect at start is a better default assumption I believe. And it will be great if JHipster could support us in this path too.

Thanks!

Tonterias commented:
> And it will be great if JHipster could support us in this path too.

Jhipster is eager to get help from programmers that use and want to improve any feature, but they will ask you if you want to do it.  They will help you, but most part will be on your fingers. May be they could pay you $100 or something like that. Sometimes they do it.

Do you want to do it? **Do you want to be Jhipster contributor?** 


PD: I'm not a programmer and they have asked me a couple times that I spotted a problem, but I could not programm at that level. I just use Jhipster and find things that I can not do or  that are broken, so I create issues and they fix it. At least, I help with that :-)

SidMorad commented:
> Do you want to do it? Do you want to be Jhipster contributor?

I don't know how hard is that, Yes :-)

I am a little bit familiar with JHipster code base, and had a [little contribution](https://github.com/jhipster/generator-jhipster/pull/5961) in the past. From technical point of view JHipster entity generator use a json file located at `.jhipster/*.json` as metadata to generate required changes. And what we are asking is like making this metadata immutable and put future changes in next versions. something like `.jhipster/Book-v1.json` `.jhipster/Book-v2.json`.  Or having still one file `.jhipster/Book.json` and using versioning inside of that file. Either ways I see this as heavy change request which may touch a lot of existing code. And that's why I didn't dare asked for this myself ;-) Having said that I might be wrong and JHipster developers could come up with a clever solution for this.

Another thing is a little bit philosophical, is JHipster a tool for starting a project? or a tool for developing a project? I think what is happening more often is people use JHipster to start a project and it is easy to reach a point that `jhipster upgrade` became impossible and they have to go their separate ways. So if JHipster have intention to be used more often during development then this type of change request may be interesting otherwise I think not.

Tonterias commented:
The first RTL! Well done! 

I have plenty of closed issues :-)  
https://github.com/jhipster/generator-jhipster/issues?q=is%3Aclosed+is%3Aissue+author%3ATonterias    

Please, let me know if I am wrong:

I would think of this as any other problem where Jhipster needs a different config (doing different things) for Dev and Prod. Most of the times they have solved those problems using the @Profile annotation and separating both configs (let say doing things differently in those environments).

Now, is there a way to make two configs for liquibase: one for Dev as it is working now and another different one for Prod?

Does it make sense to you so far? 

I have no idea about how to do it, but I can think using similarities (and write nonsense ;-)

If it makes sense, we should find the way to explain who we would do it. Once we have it clear, we can raise the issue/request.

Tonterias commented:
Hi, SidMorad. Did you see my last comment? Hope you are doing fine

Tonterias2 commented:
Checking

SidMorad commented:
Hi Mike,

> Did you see my last comment? Hope you are doing fine

Thanks, some weeks my workload is high and last week was one of them.

> If it makes sense, we should find the way to explain how we would do it. Once we have it clear, we can raise the issue/request.

I think JHipster developers do have deeper understanding of JHipster's code base, therefore there is high chance our suggested/potential solution will not be used by them. So I think it is better we just explain what the problem is and then we continue to contribute to the issue by any ways we can.

Tonterias commented:
Hi @SidMorad 

I was following the step to open a request when I've found this:

https://github.com/jhipster/generator-jhipster/issues/11398

Please take a look!

https://github.com/jhipster/generator-jhipster/pull/11419

SidMorad commented:
> jhipster/generator-jhipster#11419

Yes, this is exactly what we want.

I like `jhipster --incremental-changelog` flag, good naming!
